NetscapeBookmarkParser.php

```diff
@@ -0,0 +1,222 @@
+                }
+
+                if (preg_match('/add_date="(.*?)"/i', $line, $m8)) {
+                    $this->items[$i]['time'] = $this->parseDate($m8[1]);
+                } else {
+                    $this->items[$i]['time'] = time();
+                }
+
+                if (preg_match('/(public|published|pub)="(.*?)"/i', $line, $m9)) {
+                    $this->items[$i]['pub'] = $this->parseBoolean($m9[2], false) ? 1 : 0;
+                } elseif (preg_match('/(private|shared)="(.*?)"/i', $line, $m10)) {
+                    $this->items[$i]['pub'] = $this->parseBoolean($m10[2], true) ? 0 : 1;
+                }
+
+                $i++;
+            }
+        }
+        ksort($this->items);
+        return $this->items;
+    }
+
+
+    /**
+     * Parses a formatted date
+     *
+     * @see http://php.net/manual/en/datetime.formats.compound.php
+     * @see http://php.net/manual/en/function.strtotime.php
+     *
+     * @param string $date formatted date
+     *
+     * @return int Unix timestamp corresponding to a successfully parsed date,
+     *             else current date and time
+     */
+    public static function parseDate($date)
+    {
+        if (strtotime('@'.$date)) {
+            // Unix timestamp
+            return strtotime('@'.$date);
+        } else if (strtotime($date)) {
+            // attempt to parse a known compound date/time format
+            return strtotime($date);
+        }
+        // current date & time
+        return time();
+    }
+
+    /**
+     * Parses the value of a supposedly boolean attribute
+     *
+     * @param string $value   Attribute value to evaluate
+     *
+     * @return mixed 'true' when the value is evaluated as true
+     *               'false' when the value is evaluated as false
+     *               $this->defaultPub if the value is not a boolean
+     */
+    public function parseBoolean($value) {
+        if (! $value) {
+            return false;
+        }
+        if (! is_string($value)) {
+            return true;
+        }
+
+        if (preg_match("/^(".self::TRUE_PATTERN.")$/i", $value)) {
+            return true;
+        }
+        if (preg_match("/^(".self::FALSE_PATTERN.")$/i", $value)) {
+            return false;
+        }
+        return $this->defaultPub;
+    }
+
+    /**
+     * Sanitizes the content of a string containing Netscape bookmarks
+     *
+     * This removes extra newlines, trailing spaces and tabs.
+     *
+     * @param string $bookmarkString Original bookmark string
+     *
+     * @return string Sanitized bookmark string
+     */
+    public static function sanitizeString($bookmarkString)
+    {
+        $sanitized = $bookmarkString;
+        $sanitized = str_replace(array("\r", "\t"), array('',' '), $sanitized);
+
+        $sanitized = preg_replace_callback(
+            '@<DD>(.*?)(<A|<\/|<DL|<DT|<P)@mis',
+            function($match) {
+                return '<DD>'.str_replace("\n", '<br>', trim($match[1])).'</';
```

Comment:
There's a bit of magic involved here, as well as in the main loop, with a mix of single- and multi-line regex flags. What this method (supposedly) does is converting all data related to a given link so it is represented in a single line, with `<br>` for line breaks, hence the main parsing loop iterating on the resulting "lines".

It feels like the sanitizing/trimming ops could be further simplified to gain in readability.